File: leta/daemon/handlers/show.py
"""Handler for show command."""

from pathlib import Path
from typing import Union

from ..rpc import ShowParams
from ...lsp.types import (
    TextDocumentPositionParams,
    DocumentSymbolParams,
    TextDocumentIdentifier,
    Position,
)
from ...utils.text import read_file_content, get_lines_around
from .base import HandlerContext, find_symbol_at_line, expand_variable_range, LocationDict


async def handle_show(
    ctx: HandlerContext, params: ShowParams
) -> Union[list[LocationDict], dict[str, object]]:
    """Handle show command, returning either locations or definition content."""
    body = params.body

    if params.direct_location:
        return await _handle_direct_definition(ctx, params, body)

    if body:
        return await _handle_definition_body(ctx, params)
    
    return await _handle_location_request(ctx, params)


async def _handle_direct_definition(
    ctx: HandlerContext, params: ShowParams, body: bool
) -> Union[list[LocationDict], dict[str, object]]:
    path = Path(params.path).resolve()
    workspace_root = Path(params.workspace_root).resolve()
    line = params.line
    context = params.context
    head = params.head or 200
    symbol_name = params.symbol_name
    symbol_kind = params.symbol_kind

    rel_path = ctx.relative_path(path, workspace_root)
    content = read_file_content(path)
    lines = content.splitlines()

    if body:
        range_start = params.range_start_line
        range_end = params.range_end_line

        if range_start is not None and range_end is not None:
            start = range_start - 1
            end = range_end - 1

            if start == end and symbol_kind in ("Constant", "Variable"):
                end = expand_variable_range(lines, start)
        else:
            workspace = await ctx.session.get_or_create_workspace(path, workspace_root)
            doc = await workspace.ensure_document_open(path)
            assert workspace.client is not None
            result = await workspace.client.send_request(
                "textDocument/documentSymbol",
                DocumentSymbolParams(textDocument=TextDocumentIdentifier(uri=doc.uri)),
            )
            if result:
                symbol = find_symbol_at_line(result, line - 1)
                if symbol:
                    start = symbol["range_start"]
                    end = symbol["range_end"]
                else:
                    start = end = line - 1
            else:
                start = end = line - 1

        if context > 0:
            start = max(0, start - context)
            end = min(len(lines) - 1, end + context)

        total_lines = end - start + 1
        truncated = total_lines > head
        if truncated:
            end = start + head - 1

        return {
            "path": rel_path,
            "start_line": start + 1,
            "end_line": end + 1,
            "content": "\n".join(lines[start : end + 1]),
            "truncated": truncated,
            "total_lines": total_lines,
            "head": head,
            "symbol": symbol_name,
        }
    else:
        location: LocationDict = {
            "path": rel_path,
            "line": line,
            "column": params.column,
        }

        if context > 0 and path.exists():
            ctx_lines, start, end = get_lines_around(content, line - 1, context)
            location["context_lines"] = ctx_lines
            location["context_start"] = start + 1

        return [location]


async def _handle_definition_body(ctx: HandlerContext, params: ShowParams) -> dict[str, object]:
    locations = await _handle_location_request(ctx, params)
    if not locations:
        return {"error": "Definition not found"}

    loc = locations[0]
    workspace_root = Path(params.workspace_root).resolve()
    path_val = loc.get("path")
    line_val = loc.get("line")
    assert path_val is not None and line_val is not None
    file_path = workspace_root / path_val
    target_line = line_val - 1
    context = params.context
    head = params.head or 200
    symbol_name = params.symbol_name

    workspace, doc, _ = await ctx.get_workspace_and_document({
        "path": str(file_path),
        "workspace_root": params.workspace_root,
    })

    assert workspace.client is not None

    result = await workspace.client.send_request(
        "textDocument/documentSymbol",
        DocumentSymbolParams(textDocument=TextDocumentIdentifier(uri=doc.uri)),
    )

    content = read_file_content(file_path)
    lines = content.splitlines()

    if result:
        symbol = find_symbol_at_line(result, target_line)
        if symbol:
            start = symbol["range_start"]
            end = symbol["range_end"]
            if context > 0:
                start = max(0, start - context)
                end = min(len(lines) - 1, end + context)

            total_lines = end - start + 1
            truncated = total_lines > head
            if truncated:
                end = start + head - 1

            return {
                "path": path_val,
                "start_line": start + 1,
                "end_line": end + 1,
                "content": "\n".join(lines[start : end + 1]),
                "truncated": truncated,
                "total_lines": total_lines,
                "head": head,
                "symbol": symbol_name,
            }
        else:
            return {"error": "Language server does not provide symbol ranges"}

    return {
        "path": path_val,
        "start_line": line_val,
        "end_line": line_val,
        "content": lines[target_line] if target_line < len(lines) else "",
    }


async def _handle_location_request(
    ctx: HandlerContext, params: ShowParams
) -> list[LocationDict]:
    workspace, doc, _ = await ctx.get_workspace_and_document({
        "path": params.path,
        "workspace_root": params.workspace_root,
    })
    line, column = ctx.parse_position({"line": params.line, "column": params.column})
    context = params.context

    assert workspace.client is not None

    result = await workspace.client.send_request(
        "textDocument/definition",
        TextDocumentPositionParams(
            textDocument=TextDocumentIdentifier(uri=doc.uri),
            position=Position(line=line, character=column),
        ),
    )

    return ctx.format_locations(result, workspace.root, context)

File: crates/leta-daemon/src/handlers/show.rs
use std::path::PathBuf;

use leta_fs::{path_to_uri, read_file_content};
use leta_lsp::lsp_types::{DocumentSymbol, DocumentSymbolResponse, SymbolInformation};
use serde_json::{json, Value};

use super::{relative_path, HandlerContext};

pub async fn handle_show(ctx: &HandlerContext, params: Value) -> Result<Value, String> {
    let workspace_root = PathBuf::from(
        params.get("workspace_root")
            .and_then(|v| v.as_str())
            .ok_or("Missing workspace_root")?
    );
    let path = PathBuf::from(
        params.get("path")
            .and_then(|v| v.as_str())
            .ok_or("Missing path")?
    );
    let line = params.get("line")
        .and_then(|v| v.as_u64())
        .ok_or("Missing line")? as u32;
    let context = params.get("context")
        .and_then(|v| v.as_u64())
        .unwrap_or(0) as u32;
    let head = params.get("head")
        .and_then(|v| v.as_u64())
        .unwrap_or(200) as u32;
    let symbol_name = params.get("symbol")
        .and_then(|v| v.as_str())
        .map(String::from);
    let symbol_kind = params.get("kind")
        .and_then(|v| v.as_str())
        .map(String::from);
    let range_start_line = params.get("range_start_line")
        .and_then(|v| v.as_u64())
        .map(|v| v as u32);
    let range_end_line = params.get("range_end_line")
        .and_then(|v| v.as_u64())
        .map(|v| v as u32);
    let _direct_location = params.get("direct_location")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);

    let path = path.canonicalize().unwrap_or(path);
    let rel_path = relative_path(&path, &workspace_root);

    let content = read_file_content(&path).map_err(|e| format!("Failed to read file: {}", e))?;
    let lines: Vec<&str> = content.lines().collect();

    let (start, end) = if let (Some(range_start), Some(range_end)) = (range_start_line, range_end_line) {
        let start = (range_start as usize).saturating_sub(1);
        let mut end = (range_end as usize).saturating_sub(1);

        if start == end {
            if let Some(ref kind) = symbol_kind {
                if kind == "Constant" || kind == "Variable" {
                    end = expand_variable_range(&lines, start);
                }
            }
        }

        (start, end)
    } else {
        ctx.session.get_or_create_workspace(&path, &workspace_root).await?;
        let client = ctx.session.get_workspace_client(&path, &workspace_root).await
            .ok_or("Failed to get LSP client")?;

        let uri = path_to_uri(&path);
        let symbol_params = json!({
            "textDocument": {"uri": uri}
        });

        let result: Result<DocumentSymbolResponse, _> = client
            .send_request("textDocument/documentSymbol", symbol_params)
            .await;

        match result {
            Ok(response) => {
                if let Some((s, e)) = find_symbol_range(&response, (line as usize).saturating_sub(1)) {
                    (s, e)
                } else {
                    let l = (line as usize).saturating_sub(1);
                    (l, l)
                }
            }
            Err(_) => {
                let l = (line as usize).saturating_sub(1);
                (l, l)
            }
        }
    };

    let (start, end) = if context > 0 {
        let start = start.saturating_sub(context as usize);
        let end = (end + context as usize).min(lines.len().saturating_sub(1));
        (start, end)
    } else {
        (start, end)
    };

    let total_lines = (end - start + 1) as u32;
    let truncated = total_lines > head;
    let end = if truncated {
        start + (head as usize) - 1
    } else {
        end
    };

    let content_slice: Vec<&str> = lines[start..=end.min(lines.len().saturating_sub(1))].to_vec();

    Ok(json!({
        "path": rel_path,
        "start_line": start + 1,
        "end_line": end + 1,
        "content": content_slice.join("\n"),
        "truncated": truncated,
        "total_lines": total_lines,
        "symbol": symbol_name,
    }))
}

fn find_symbol_range(response: &DocumentSymbolResponse, target_line: usize) -> Option<(usize, usize)> {
    match response {
        DocumentSymbolResponse::Nested(symbols) => find_in_document_symbols(symbols, target_line),
        DocumentSymbolResponse::Flat(symbols) => find_in_symbol_information(symbols, target_line),
    }
}

fn find_in_document_symbols(symbols: &[DocumentSymbol], target_line: usize) -> Option<(usize, usize)> {
    for sym in symbols {
        let start = sym.range.start.line as usize;
        let end = sym.range.end.line as usize;

        if start <= target_line && target_line <= end {
            if !sym.children.is_empty() {
                if let Some(child_range) = find_in_document_symbols(&sym.children, target_line) {
                    return Some(child_range);
                }
            }
            return Some((start, end));
        }
    }
    None
}

fn find_in_symbol_information(symbols: &[SymbolInformation], target_line: usize) -> Option<(usize, usize)> {
    for sym in symbols {
        let line = sym.location.range.start.line as usize;
        if line == target_line {
            return Some((line, sym.location.range.end.line as usize));
        }
    }
    None
}

fn expand_variable_range(lines: &[&str], start_line: usize) -> usize {
    if start_line >= lines.len() {
        return start_line;
    }

    let first_line = lines[start_line];

    let mut open_parens = first_line.matches('(').count() as i32 - first_line.matches(')').count() as i32;
    let mut open_brackets = first_line.matches('[').count() as i32 - first_line.matches(']').count() as i32;
    let mut open_braces = first_line.matches('{').count() as i32 - first_line.matches('}').count() as i32;

    let triple_double = first_line.matches("\"\"\"").count();
    let triple_single = first_line.matches("'''").count();
    let mut in_multiline_string = triple_double % 2 == 1 || triple_single % 2 == 1;

    if open_parens == 0 && open_brackets == 0 && open_braces == 0 && !in_multiline_string {
        return start_line;
    }

    for i in (start_line + 1)..lines.len() {
        let line = lines[i];

        if in_multiline_string {
            if line.contains("\"\"\"") || line.contains("'''") {
                in_multiline_string = false;
                if open_parens == 0 && open_brackets == 0 && open_braces == 0 {
                    return i;
                }
            }
            continue;
        }

        open_parens += line.matches('(').count() as i32 - line.matches(')').count() as i32;
        open_brackets += line.matches('[').count() as i32 - line.matches(']').count() as i32;
        open_braces += line.matches('{').count() as i32 - line.matches('}').count() as i32;

        let triple_double = line.matches("\"\"\"").count();
        let triple_single = line.matches("'''").count();
        if triple_double % 2 == 1 || triple_single % 2 == 1 {
            in_multiline_string = true;
            continue;
        }

        if open_parens <= 0 && open_brackets <= 0 && open_braces <= 0 {
            return i;
        }
    }

    start_line
}
